nntrainer/graph/network_graph.cpp

```diff
@@ -576,8 +576,22 @@ void NetworkGraph::inPlaceOptimize(const std::string &layer_type,
         continue;
 
       /** Share tensor with next layer */
-      l->net_input[0] = l->net_hidden[0];
-      prev_layer->net_hidden[loc] = l->net_hidden[0];
+      l->net_input[0] = l->net_hidden[0]; // I2 = O2
+      if (l->getType() == BatchNormalizationLayer::type) {
+        prev_layer->net_hidden[loc] = l->net_hidden[0]; // O1 = O2
+      } else if (l->getType() == ActivationLayer::type) {
+        // THIS IS NOT WORKING BECAUSE INITIALIZES MAKES NEW SHARED_PTR OF TENSOR
+        // THAN KEEPING THE SAME SHARED_PTR AND JUST CHANGING THE TENSOR IN IT
+        prev_layer->net_hidden[loc]->getGradientRef() =
+          l->net_hidden[0]->getVariableRef(); // O1.G = O2.V
+        prev_layer->net_hidden[loc]->getVariableRef() =
+          l->net_hidden[0]->getVariableRef(); // O1.V = O2.V
+      }
```

Comment:
is it right to point grad and var to 02.v?

Reply:
Yes, this is intended. The derivative optimization of different layers happens differently.
For layers working in-place, like activation, these layers share derivative from common shared memory.
However, for layers that are not in-place, these layers use their output memory as the derivative memory.